This patch adds a button on the toolbar to change the sort function.

Updated for 1.3.3 CVS 2002-07-30 by Marcin Juszkiewicz <marcin@amiga.pl>

Index:rox-2002-05-26-sortbutton/ROX-Filer/src/toolbar.c
*** rox-2002-05-26/ROX-Filer/src/toolbar.c	Mon May 27 09:46:34 2002
--- rox-2002-05-26-sortbutton/ROX-Filer/src/toolbar.c	Mon May 27 10:51:47 2002
***************
*** 81,86 ****
--- 81,88 ----
  				    FilerWindow *filer_window);
  static void toolbar_hidden_clicked(GtkWidget *widget,
  				   FilerWindow *filer_window);
+ static void toolbar_sort_clicked(GtkWidget *widget,
+ 				   FilerWindow *filer_window);
  static GtkWidget *add_button(GtkWidget *bar, Tool *tool,
  				FilerWindow *filer_window);
  static GtkWidget *create_toolbar(FilerWindow *filer_window);
***************
*** 124,129 ****
--- 126,135 ----
  	 toolbar_size_clicked, DROP_NONE, TRUE,
  	 NULL},
  	
+ 	{N_("Sort"), GTK_STOCK_SORT_ASCENDING, N_("Change sort criteria"),
+ 	 toolbar_sort_clicked, DROP_NONE, TRUE,
+ 	 NULL},
+ 	
  	{N_("Details"), GTK_STOCK_JUSTIFY_LEFT, N_("Show extra details"),
  	 toolbar_details_clicked, DROP_NONE, TRUE,
  	 NULL},
***************
*** 317,322 ****
--- 323,361 ----
  		bev->type == GDK_BUTTON_RELEASE && bev->button == 1);
  }
  
+ /* This typedef makes things a lot easier... */
+ typedef int (*sort_fn_t)(const void *a, const void *b);
+ 
+ static void toolbar_sort_clicked(GtkWidget *widget, FilerWindow *filer_window)
+ {
+ 	GdkEventButton	*bev;
+ 	int i, current, next;
+ 	gboolean adjust;
+ 
+ 	static sort_fn_t sort_fns[]={
+ 	  sort_by_name, sort_by_type, sort_by_date, sort_by_size
+ 	};
+ #define NSORT (sizeof(sort_fns)/sizeof(sort_fn_t))
+ 
+ 	bev = (GdkEventButton *) gtk_get_current_event();
+ 	adjust=(bev->button == 2) && bev->type == GDK_BUTTON_RELEASE;
+ 
+ 	current=-1;
+ 	for(i=0; i<NSORT; i++)
+ 	  if(filer_window->sort_fn==sort_fns[i]) {
+ 	    current=i;
+ 	    break;
+ 	  }
+ 	if(current==-1)
+ 	  next=0;
+ 	else if(adjust)
+ 	  next=(current+NSORT-1)%NSORT;
+ 	else
+ 	  next=(current+1)%NSORT;
+ 	  
+ 	display_set_sort_fn(filer_window, sort_fns[next]);
+ }
+ 
  static void toolbar_details_clicked(GtkWidget *widget,
  				    FilerWindow *filer_window)
  {
